src/mlpack/tests/hmm_test.cpp

```diff
@@ -835,8 +835,39 @@ BOOST_AUTO_TEST_CASE(GaussianHMMPredictTest)
       }
   };
 
+  const double loglikelihoodRef = -2734.43;
+
+  {
+    auto loglikelihood = hmm.LogLikelihood(obs);
+    BOOST_REQUIRE_CLOSE(loglikelihood, loglikelihoodRef, 1e-3);
+  }
+
+  {
+    double loglikelihood;
+    arma::vec prevForwardLogProb;
+    arma::vec forwardLogProb;
+    for (size_t t = 0; t<obs.n_cols; ++t)
+    {
+      loglikelihood = hmm.LogLikelihood(obs.col(t), loglikelihood,
+                                        prevForwardLogProb, forwardLogProb);
+    }
+    BOOST_REQUIRE_CLOSE(loglikelihood, loglikelihoodRef, 1e-3);
+  }
+
+  {
+    double loglikelihood = 0;
+    arma::vec prevForwardLogProb;
+    arma::vec forwardLogProb;
+    for (size_t t = 0; t<obs.n_cols; ++t)
+    {
+      auto logScale = hmm.LogScale(obs.col(t), prevForwardLogProb,
+                                    forwardLogProb);
+      loglikelihood += logScale;
+    }
+    BOOST_REQUIRE_CLOSE(loglikelihood, loglikelihoodRef, 1e-3);
+  }
```

Comment:
These are testing different variations of loglikelihood calculation of the same hmm object. That's why I put them in one test case otherwise I had to repeat the hmm construction part several times for each individual test case which ends up in a lot of redundancy in hmm test unit which makes the test code difficult to maintain and error prone in future.
I tried to add comments inside the test case to explain each section.

Reply:
Understood---thanks for the clarification and the comments! :+1: